### PR TITLE
Fix ``not-callable`` for attributes that alias ``NamedTuple``

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Closes part of #1730, specifically this comment: https://github.com/PyCQA/pylint/issues/1730#issuecomment-489572303

I tried to destructure the `if` a little bit because I got a `too-many-boolean-expressions` warning and it became really unreadable.. This did lead to some code duplication though..

### CODE REVIEW

File: pylint/checkers/typecheck.py [hunk 1]
```diff
@@ -1294,16 +1294,23 @@ def visit_call(self, node: nodes.Call) -> None:
         # Ignore instances of descriptors since astroid cannot properly handle them
         # yet
         if called and not called.callable():
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/typecheck.py:1296]: Maybe we can create a function from line 1296 to line 1318 and use premature return to simplify this expression ?

```python
 if not isinstance(called, astroid.Instance):
    self._check_uninferable_call(node)
    return
if not has_known_bases(called):
    return
....        
```

**Review by Pierre-Sassoulas — COMMENTED** [review]
I think this would benefit from a refactor even if that's a little outside of the scope of the current MR. Let me know what you think :)

**Comment by DanielNoord** [comment]
Makes sense! I'll get to it tomorrow 😄

File: pylint/checkers/typecheck.py [hunk 2]
```diff
@@ -1567,6 +1551,34 @@ def _check_invalid_sequence_index(self, subscript: nodes.Subscript):
...
+                    # Ignore descriptor instances
+                    if "__get__" in inferred_call.locals:
+                        return
+                    # NamedTuple instances are callable
+                    if inferred_call.qname() == "typing.NamedTuple":
+                        return
+
+            self.add_message("not-callable", node=node, args=node.func.as_string())
+        else:
+            self._check_uninferable_call(node)
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/typecheck.py:1580]:
```suggestion
if not inferred_call or inferred_call.callable():
    return self._check_uninferable_call(node)
if not isinstance(inferred_call, astroid.Instance):
    return
# Don't emit if we can't make sure this object is callable.
if not has_known_bases(inferred_call):
    return
if inferred_call.parent and isinstance(
        inferred_call.scope(), nodes.ClassDef
):
    # Ignore descriptor instances
    if "__get__" in inferred_call.locals:
        return
    # NamedTuple instances are callable
    if inferred_call.qname() == "typing.NamedTuple":
        return
self.add_message("not-callable", node=node, args=node.func.as_string())
```
One step further :)

Comment (DanielNoord) [reply pylint/checkers/typecheck.py]: Slightly adapted this suggestion and added some extra white lines + comments in latest commit.

File: pylint/checkers/typecheck.py [hunk 3]
```diff
@@ -1559,25 +1559,28 @@ def _check_not_callable(
...
-
-            self.add_message("not-callable", node=node, args=node.func.as_string())
-        else:
+        # Handle uninferrable calls
+        if not inferred_call or inferred_call.callable():
             self._check_uninferable_call(node)
+            return
+
+        # Ignore non instances
+        if not isinstance(inferred_call, astroid.Instance):
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/typecheck.py:1568]: Sorry I was wrong in my first proposition I edited it. I think there's a guaranteed message for non instances. (The add_message is in the same scope)

Comment (DanielNoord) [reply pylint/checkers/typecheck.py:1568]: This doesn't fail any tests.. The coveralls comments seems broken, but are we actually testing this?

Comment (DanielNoord) [reply pylint/checkers/typecheck.py:1568]: Relevant commit:
https://github.com/PyCQA/pylint/commit/ea14cad3187aa2efdd0ae7b520b2ba44797c0863

Seems like we can remove it? All `inferred_call`'s that we test are `Instance`'s

Comment (Pierre-Sassoulas) [reply pylint/checkers/typecheck.py:1568]: I think you're right, but it's really hard to reconstruct code example that would trigger a particular piece of code though. (Or "prove" that it's impossible). I hope primer++ will fix this.

**Review by Pierre-Sassoulas — APPROVED** [review]
I approve with our without the instance test, you decide :smile:

**Comment by DanielNoord** [comment]
I've removed it for now. `called` is the inference of `nodes.Call.func`, seems like that should always be an `Instance` (considering `nodes.Const` subclasses `Instance` as well). The original commit also doesn't provide any good reasons to think otherwise, I feel the `if` was added simply because the issue referred to instances originally.

**Comment by DanielNoord** [comment]
Thank our test suite: this change was not trivial. Reverting and adding the message back in, it should actually be called see:
https://github.com/PyCQA/pylint/blob/566a377154ce68e795aa3f952bb70bd987045228/tests/functional/d/deprecated/deprecated_methods_py38.py#L30

**Pull request merged**